### DISCUSSION TITLE
Validation fails for polish ePUAP trusted signature

### DISCUSSION THREAD

**Discussion by janprzychodniak** [MatthiasValvekens/pyHanko#318] [Q&A]
I'm trying to validate the signature in the following document:
[Skierowanie_na_wstepne_badania_lekarskie_Testowy_Janusz_signed_EPUAP.pdf](https://github.com/MatthiasValvekens/pyHanko/files/12595505/Skierowanie_na_wstepne_badania_lekarskie_Testowy_Janusz_signed_EPUAP.pdf)

To my knowledge, the signature is valid (based on Adobe Acrobat validation, see image below).
![image](https://github.com/MatthiasValvekens/pyHanko/assets/117818900/f407694e-2104-422d-b300-060646f7768a)

However pyHanko says that "The signature is cryptographically unsound.".

Here's my code:
```python
def get_document_sign_info(pdf_document:BinaryIO) -> DocumentSignatureInfo:
    try:
        r = PdfFileReader(pdf_document, strict=False)

        doc = DocumentSignatureInfo()
        signatures:list[EmbeddedPdfSignature] = r.embedded_signatures
        if len(signatures) == 0:
            return doc
        
        # Signatures are ordered form oldest to newest
        # Only validate newest (otherwise docmdp_ok will
        # be invalid)
        signature = signatures[-1]

        validate_result = validate_pdf_signature(signature, skip_diff=True)
        print(validate_result.pretty_print_details())
```

And here's the result from pretty_print_details() method:

```
Signer info
-----------
Certificate subject: "Organization: Kancelaria Prezesa Rady Ministrow, Common Name: Minister do spraw informatyzacji - pieczec podpisu zaufanego, Organization Identifier: VATPL-5261645000, Country: PL"
Certificate SHA1 fingerprint: f1ea5832e7358ffc9d9b9afe040b1f01a0541a46
Certificate SHA256 fingerprint: f13cf1de71f57181b13724467a071f322e9e83032418c00eab1fc84cd9d36d0f
Trust anchor: "No path to trust anchor found."
The signer's certificate is untrusted.


Integrity
---------
The signature is cryptographically unsound.

The digest algorithm used was 'sha256'.
The signature mechanism used was 'sha256_rsa'.


Signing time
------------
Signing time as reported by signer: 2023-09-12T09:52:12+02:00


Modifications
-------------
The signature covers the entire file.


Bottom line
-----------
The signature is judged INVALID.
```

**Note:** for my use case the certificate does not have to be trusted, at least at this stage of the project. The signature must only be valid, intact and docmdp_ok.

---

So bottom line is: What am I doing wrong?

**Comment by ofcaah** [comment]
Matthias, please note, that it's the same (or similar) issue that I was mailing you privately some time ago. There's possibly also some specific way to determine in who's name signature was created, as Adobe Acrobat states, that document has been correctly signed by Tomasz Drzewiecki:

![image](https://github.com/MatthiasValvekens/pyHanko/assets/6192229/2d03d7c3-25de-4542-93f6-e57e556c8de0)

**Comment by MatthiasValvekens** [comment]
~OK, I'm not _entirely_ sure why this fails, but I investigated and I think I have a lead.~ EDIT: On further inspection (read answers below), I'm now quite confident that this is indeed the root cause.

First, I reduced the validation failure down to the RSA signature validation itself failing. The weird thing is that not all PDF tools agree:

 - `pdfsig` from poppler-utils considers the signature invalid
 - Acrobat accepts it
 - iText (Java; OpenJDK 20 with the Sun security provider) accepts it

I've stepped through both the iText validation procedure and pyHanko's with a debugger, and I've verified that all the input parameters fed into the backend RSA library are identical. So the discrepancy lies somewhere deeper.

Debugging `pyca/cryptography`'s calls to openssl is harder (the error is raised from within OpenSSL), but on the iText/Java side I could go deeper---the validation in the Sun security provided is implemented in plain old Java, so easier to debug. I found the following anomaly: after the unpadding step in the validation algorithm, the returned data (in hex) is as follows:

```
302f300b06096086480165030402010420cf55a54c74490a158e1c3a6bea104b739f88ca243f75f6cca0eb9a5063cc06d7
```

Decoding the DER, we get this: https://lapo.it/asn1js/#MC8wCwYJYIZIAWUDBAIBBCDPVaVMdEkKFY4cOmvqEEtzn4jKJD919syg65pQY8wG1w. But this violates PKCS#1 in a subtle way: the `parameters` field in the algorithm identifier part is missing (it must be present and set to `NULL`). The Java implementation (at least in my JDK) ignores this detail when validating, but perhaps OpenSSL doesn't... To be further investigated.


TL;DR the signature encoding is invalid for a pedantic technical reason, which could be the reason why OpenSSL refuses to accept it

**Reply by MatthiasValvekens** [reply]
To add to the above: the OpenJDK sources contain the following relevant comment:

```
            byte[] digest = getDigestValue();
            byte[] decrypted = RSACore.rsa(sigBytes, publicKey);
            byte[] unpadded = padding.unpad(decrypted);
            // https://www.rfc-editor.org/rfc/rfc8017.html#section-8.2.2
            // Step 4 suggests comparing the encoded message instead of the
            // decoded, but some vendors might omit the NULL params in
            // digest algorithm identifier.
            byte[] decodedDigest = RSAUtil.decodeSignature(digestOID, unpadded);
            return MessageDigest.isEqual(digest, decodedDigest);
```

In other words, if OpenSSL implements the spec as written (comparing the encoded values instead of the decoded content, see step 4 of https://datatracker.ietf.org/doc/html/rfc8017#section-8.2.2), then this validation failure is entirely expected.

Of course, I didn't (yet) verify whether this is _really_ what OpenSSL does, but if that's what's going on, it's not something I can easily override in pyHanko...

**Reply by MatthiasValvekens** [reply]
There appear to be some differences between OpenSSL flavours in exactly how this is handled, but the `master` branch in "mainline" OpenSSL currently has this logic in its RSA verification routine: https://github.com/openssl/openssl/blob/388a8e731445d190a46ec27b2ff5b4bf334d526b/crypto/rsa/rsa_sign.c#L421-L428

In other words, OpenSSL does approximately the following when validating an RSASSA-PKCS1v1.5 signature (assuming the applicable digest function is known):

1. Unwrap the signature using the public key
2. Validate and remove the padding from the result of step 1
3. Extract the digest from the resulting data by taking the last `digestLen` bytes of the output of step 2
4. Reconstruct the expected DER-encoded `DigestInfo` object from the output of step 3 and ensure it matches the output of step 2.
5. Compare the output of step 3 to the digest of the signed attributes.

Since the ePUAP signer that produced that file apparently uses an invalid `DigestInfo` representation when creating the signature, step 4 fails. This will be the case for all PDF software that relies on OpenSSL under the hood for signature validation...

**Reply by janprzychodniak** [reply]
Is there anything I can do to help with your research? Maybe provide more files or details?

**Reply by MatthiasValvekens** [reply]
Well, the cause has been determined as far as I'm concerned :). The signature's structure is technically not valid, and OpenSSL contains an explicit check to verify PKCS#1 compliance, but other implementations don't.

Short of allowing swappable crypto backends for validation, I'm not sure what I can do about it in pyHanko, though... The entry point I'm calling to trigger validation in pyHanko is one exposed by `cryptography`, and two or three levels of abstraction above the OpenSSL snippet I linked above.

What do you think?

**Reply by janprzychodniak** [reply]
I think that considering the signature is technically invalid, it'll be worth reporting this to ePUAP maintainers. For starters at least. I'll take care of it and I will come back to you with their response. It might take a while though, government bodies aren't exactly quick at responding. ;)

**Reply by MatthiasValvekens** [reply]
Thanks!

I've had mixed success with that in Belgium. Sure, they can be slow, but I found that it usually helps if you bypass the standard support channels by contacting the people/addresses listed in their certification policy documents directly... No "regular" user is ever going to read those 😛

**Comment by ofcaah** [comment]
On a side note, getting this fixed by gov body might prove next to impossible, as the "official" validator doesn't care and says everything's fine:
https://ec.europa.eu/digital-building-blocks/DSS/webapp-demo/validation

**Reply by MatthiasValvekens** [reply]
FWIW: The eSig-DSS validator is implemented in Java and (as far as I know) uses the default cryptographic primitives in the standard library, which means that it is tolerant of these issues for the exact same reason as iText.

**Comment by ofcaah** [comment]
We are seeing quite a lot of those signatures, and while I can whitelist SHA256 hash of gov body's certificate, this doesn't seem like a good solution -- or maybe it is?

Assuming they won't care about fixing this, which is IMO extremely probable, what other options, besides patching openssl, are on the table?

**Reply by MatthiasValvekens** [reply]
Hi @ofcaah,

To be perfectly honest with you: if you absolutely _must_ support this, then your best bet is monkeypatching `pyhanko.sign.validation.utils.validate_raw` and supplying a different RSA validation implementation that ignores this type of issue (I'm sure there are plenty). Of course this comes with all sorts of caveats, but that's the way to do it. In `pyhanko-certvalidator`, `pyhanko_certvalidator.util.validate_sig` fulfills more or less the same function.

On general principle I won't ship any such change in the core library---that would be grossly irresponsible---but if your use case requires it, you do you.

------

That said, before you do any of that: have you actually asked your government IT services what's up? Certification practice statements must have contact information---that's probably a good place to start. If they don't respond, you can always escalate your complaint to your country's TLSO. They're supposed to enforce standards compliance in eIDAS-related matters. If those people don't respond either you could try writing directly to the EU trusted list administration at EC-TL-Service@ec.europa.eu (fair warning: I've never actually tried that myself).

Also: what the "official" validator on the European Commission's website says doesn't really matter. There is no reference implementation, only the text of the specification(s) matter. The ePUAP implementation is obviously violating the provisions of ETSI TS 199 312, clause 6.2.1, which, barring any overrides in national legislation, require all RSA AdES signatures to follow PKCS#1.

**Reply by ofcaah** [reply]
OK, you definitely have a point there; lets try to fix things, before breaking more of them ;)
@janprzychodniak did you get any response to your e-mails?

**Reply by MasterTomasz** [reply]
Finally, we received the following response from the Polish authorities – here is the translated message:

_"Signing PDF documents in the Trusted Profile is implemented using the European DSS library (https://ec.europa.eu/digital-building-blocks/sites/display/DIGITAL/Digital+Signature+Service+-++DSS). Documents signed in this way are correctly validated by the online validator (https://ec.europa.eu/digital-building-blocks/DSS/webapp-demo/validation) as well as by various applications, including Adobe Acrobat Reader. We do not plan to introduce any changes to the implementation of signature submission or verification mechanisms in the near future. The different implementation in the OpenSSL library leads to a validation error in this framework."_

**Reply by MatthiasValvekens** [reply]
The last sentence surprises me... If I read it correctly, it appears to say that they tried it with OpenSSL and the result didn't validate with eSig DSS, which would be very strange.

At the same time, I think eSig DSS actually has nothing to do with this problem, otherwise it would already have been reported a long time ago. If they're loading a JCA security provider into their JVM with a broken RSA implementation, then eSig DSS would indeed output "wrong" signatures, since eSig DSS almost surely doesn't implement RSA by itself (full disclosure: I didn't actually verify this, but if you want I can find out how they delegate it).

In fact, if they're also validating the OpenSSL output with the same broken security provider, then perhaps that explains both the observations in this thread and their test results with OpenSSL.

In short, I don't find the ePUAP team's argument all that convincing... ;)

PS: reading between the lines, it looks like they're generating signatures with software-managed keys instead of using an HSM. If true, that alone is already somewhat eyebrow-raising :). In fact, I'm pretty sure that they'd never have had this issue if they would've done the signing in hardware. AFAIK most (if not all) HSMs support RSA output in PKCS#1, and I'd be very surprised if an HSM that gets PKCS#1 wrong could pass certification ;).

**Reply by MatthiasValvekens** [reply]
> AFAIK most (if not all) HSMs support RSA output in PKCS#1, and I'd be very surprised if an HSM that gets PKCS#1 wrong could pass certification ;).

Turns out I put too much stock in the certification process. Not only did come across a signature coming from a QSCD with this exact problem, it was the signature certifying _the trusted list of an entire EU member state_: #643.
